File: bytopic/divideandconquer/java/group3/MergeKSortedLists.java
//  Given an array of k linked lists that are sorted in ascending order, merge all the linked lists
// into one sorted linked list and return it.
// Sample 1:
// Input lists = [[11, 14, 16],[11, 12, 18],[13,15]]
// Output: [11,12,13,14,15,16,18]
//
// Tag: 111/150
// Tag: 23/2927, R145/2936 (overall frequency ranking)

class ListNode {
    int value;
    ListNode next;

    public ListNode(int value) {
        this.value = value;
    }
    public ListNode(int value, ListNode next) {
        this.value = value;
        this.next = next;
    }
}
class MergeKSortedLists {
    public static void main(String...args) {
        ListNode l1 = new ListNode(11);
        l1.next = new ListNode(14);
        l1.next.next = new ListNode(16);

        ListNode l2 = new ListNode(11);
        l2.next = new ListNode(12);
        l2.next.next = new ListNode(18);

        ListNode l3 = new ListNode(13);
        l3.next = new ListNode(15);

        ListNode[] lists = {l1, l2, l3};


        ListNode result = mergeKSortedLists(lists);
        while (result != null) {
            System.out.printf(result.value +", ");
            result = result.next;
        }
        System.out.println();

    }

    static ListNode mergeKSortedLists(ListNode[] lists) {
        if (lists == null || lists.length == 0) {
            return null;
        }
        return mergeKSortedListsHelper(lists, 0, lists.length - 1);
    }

    static ListNode mergeKSortedListsHelper(ListNode[] lists, int start, int end) {
        if (start == end) {
            return lists[start];
        }
        if (start + 1 == end) {
            return merge(lists[start], lists[end]);
        }

        int mid = start + (end - start) / 2;
        ListNode left = mergeKSortedListsHelper(lists, start, mid);
        ListNode right = mergeKSortedListsHelper(lists, mid + 1, end);
        return merge(left, right);
    }

    static ListNode merge(ListNode l1, ListNode l2) {
        ListNode temp = new ListNode(0);
        ListNode curr = temp;

        while(l1 != null && l2 != null) {
            if (l1.value < l2.value) {
                curr.next = l1;
                l1 = l1.next;
            } else {
                curr.next = l2;
                l2 = l2.next;
            }
            curr = curr.next;
        }
        curr.next = (l1 != null) ? l1 : l2;
        return temp.next;
    }
}

File: bytopic/divideandconquer/python/group3/merge_k_sorted_lists.py
"""
 Given an array of k linked lists that are sorted in ascending order, merge all the linked lists
into one sorted linked list and return it.

Sample 1:
Input lists = [[11, 14, 16],[11, 12, 18],[13,15]]
Output: [11, 11,12,13,14,15,16,18]

Tag: 111/150
Tag: 23/2927, R145/2936 (overall frequency ranking)
"""
import heapq
from typing import List, Optional


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


def merge_k_lists_1(lists: List[Optional[ListNode]]) -> Optional[ListNode]:
    h = []
    head = tail = ListNode(0)
    for i in range(len(lists)):
        heapq.heappush(h, (lists[i].val, i, lists[i]))

    while h:
        node = heapq.heappop(h)
        node = node[2]
        tail.next = node
        tail = tail.next
        if node.next:
            i += 1
            heapq.heappush(h, (node.next.val, i, node.next))

    return head.next


def merge_k_lists(lists: List[Optional[ListNode]]) -> Optional[ListNode]:
    ListNode.__eq__ = lambda self, other: self.val == other.val
    ListNode.__lt__ = lambda self, other: self.val < other.val
    h = []
    head = tail = ListNode(0)
    for i in lists:
        if i:
            heapq.heappush(h, (i.val, i))
    while h:
        node = heapq.heappop(h)[1]
        tail.next = node
        tail = tail.next
        if node.next:
            heapq.heappush(h, (node.next.val, node.next))
    return head.next


def main():
    head1 = ListNode(11)
    head1.next = ListNode(14)
    head1.next.next = ListNode(16)
    head2 = ListNode(11)
    head2.next = ListNode(12)
    head2.next.next = ListNode(18)
    head3 = ListNode(13)
    head3.next = ListNode(15)

    res = merge_k_lists_1([head1, head2, head3])
    while res is not None:
        print(res.val, end=",")
        res = res.next


if __name__ == "__main__":
    main()
